test/src/702-symlink_caching/main

```diff
@@ -0,0 +1,196 @@
+cvmfs_test_name="Symlink caching and dentry eviction: Proper eviction policy for kernel cache"
+cvmfs_test_autofs_on_startup=false
+cvmfs_test_suites="quick"
+
+TEST702_PRIVATE_MOUNT=
+TEST702_PIDS=
+CVMFS_TEST_702_OSXMOUNTPOINT=
+
+source ./src/702-symlink_caching/setup_teardown
+
+
+# Returns 0 if readlink is cached in kernel cache
+# Otherwise 1
+# If empty string (=error) return 2
+check_readlink_is_cached() {
+  local mntpnt="$1"
+  local symlink_name="$2"
+
+  readlink ${mntpnt}/${symlink_name} > /dev/null
+  readlink ${mntpnt}/${symlink_name} > /dev/null
+  local old_counter=$(sudo cvmfs_talk -p ${mntpnt}c/$CVMFS_TEST_REPO/cvmfs_io.$CVMFS_TEST_REPO internal affairs | grep readlink)
+
+  readlink ${mntpnt}/${symlink_name} > /dev/null
+  readlink ${mntpnt}/${symlink_name} > /dev/null
+  readlink ${mntpnt}/${symlink_name} > /dev/null
+  local new_counter=$(sudo cvmfs_talk -p ${mntpnt}c/$CVMFS_TEST_REPO/cvmfs_io.$CVMFS_TEST_REPO internal affairs | grep readlink)
+
+  if [ "$old_counter" = "$new_counter" ] && [ "$old_counter" != "" ] ; then
+    echo "0"
+  elif [ "$old_counter" != "$new_counter" ] && [ "$old_counter" != "" ] && [ "$new_counter" != "" ] ; then
+    echo "1"
+  else
+    echo "2"
+  fi
+}
+
+# for fuse3 check
+replace_short_symlink_with_long() {
+  local short_name="$1"
+  local long_name="$2"
+  local symlink_name="$3"
+
+  start_transaction $CVMFS_TEST_REPO || return $?
+  ln -sf ${long_name} /cvmfs/${CVMFS_TEST_REPO}/${symlink_name}
+  publish_repo ${CVMFS_TEST_REPO} || return $?
+
+  #remount
+  sudo cvmfs_talk -p ${mntpnt}c/$CVMFS_TEST_REPO/cvmfs_io.$CVMFS_TEST_REPO remount sync
+}
+
+cleanup_mount_on_top() {
+  echo "Cleaning up: umount mount-on-top"
+  sudo umount $TEST702_PRIVATE_MOUNT/testdir_mountpoint
+  if [ "x$CVMFS_TEST_702_OSXMOUNTPOINT" != "x" ]; then
+    sudo hdiutil detach $CVMFS_TEST_702_OSXMOUNTPOINT
+  fi
+}
+
+# should work both for libfuse2 and libfuse3 as
+# - libfuse2 does not evict dentries
+# - libfuse3 expires dentries instead of evicting
+mount_on_top() {
+  local mntpnt="$1"
+  echo $mntpnt
+
+
+  if running_on_osx; then
+    CVMFS_TEST_702_OSXMOUNTPOINT=$(sudo hdid -nomount ram://256000)
+    sudo newfs_hfs $CVMFS_TEST_702_OSXMOUNTPOINT || return 30
+    sudo mount -t hfs $CVMFS_TEST_702_OSXMOUNTPOINT $mntpnt/testdir_mountpoint || return 31
+    sudo mkdir $mntpnt/testdir_mountpoint/self || return 32
+  else
+    sudo mount -t proc none $mntpnt/testdir_mountpoint || return 33
+  fi
+  trap cleanup_mount_on_top EXIT HUP INT TERM
+
+  ls $mntpnt/testdir_mountpoint/self > /dev/null || return 34
+  diff $mntpnt/testdir_mountpoint/self/maps /proc/self/maps || return 35
```

Comment:
I guess this fails on macOS, so perhaps restrict this check to linux only; or remove.

Reply:
why? other directory possible to mount on top that exists on all platforms?